WIP dns resolution worker v2

bjlaub opened the pull request:
## Before this PR
<!-- What's wrong with the current state of the world and why change it now? -->

## After this PR
==COMMIT_MSG==
<!-- User-facing outcomes this PR delivers -->
==COMMIT_MSG==

## Possible downsides?
<!-- Please describe any way users could be negatively affected by this PR. -->

Commit:
WIP dns resolution worker v2

Commit:
checkstyle

Commit:
test dynamic name resolution after initial state is set

Commit:
fix test

Commit:
getInternalDialogueChannel maps from output of dns resolver thread

Commit:
stuff

carterkozak commented on dialogue-clients/src/main/java/com/palantir/dialogue/clients/DialogueDnsResolver.java:
```diff
@@ -0,0 +1,31 @@
+import com.google.common.collect.ImmutableSetMultimap;
+import java.net.InetAddress;
+
+public interface DialogueDnsResolver {
```
probably want to delete this, it should duplicate the `DialogueDnsResolver` declared in dialogue-core in the target branch

carterkozak commented on dialogue-clients/src/main/java/com/palantir/dialogue/clients/ServicesConfigBlockWithResolvedHosts.java:
```diff
@@ -0,0 +1,70 @@
+import java.util.Objects;
+
+@DoNotLog
+final class ServicesConfigBlockWithResolvedHosts {
```
It might be a bit safer to use immutables to generate these sorts of types, since it'll include all fields in equals/hashCode

carterkozak commented on dialogue-clients/src/main/java/com/palantir/dialogue/clients/ReloadingClientFactory.java:
```diff
@@ -512,4 +542,49 @@ public String toString() {
         }
     }
+
+    private static final class InternalDialogueChannelConfiguration {
```
probably safest to use immutables here as well, matching other internal types used in this module (e.g. the ChannelCacheKey/etc)

Commit:
remove duplicate DialogueDnsResolver

Commit:
use immutables for InternalDialogueChannelConfiguration

Commit:
use immutables for ServicesConfigBlockWithResolvedHosts

Commit:
fix import

bjlaub commented on dialogue-clients/src/main/java/com/palantir/dialogue/clients/ServicesConfigBlockWithResolvedHosts.java:
```diff
@@ -0,0 +1,34 @@
+@Value.Immutable
+interface ServicesConfigBlockWithResolvedHosts {
+    @Value.Parameter
+    ServicesConfigBlock scb();
```
@carterkozak thoughts on making this an `Optional`? a thing i'm concerned about is: we ideally need to set an initial value for a `SettableRefreshable` of this type (see here: https://github.com/palantir/dialogue/pull/2141/files#diff-6d1086f7aca5ef649647a82c52683f1d85aa8c3a9568e9b4510669f1e749fc89R75-R76) which will be provided before the full round of initial name resolutions has completed. We could construct an "empty" `ServicesConfigBlock` but i worry that becomes a bit error-prone if the types within change to being required at build time.

carterkozak replied:
If those types changed, we’d break a lot more than this — I don’t think it would be possible to add required types to that type.

Commit:
set initial dns resolution result to empty

Commit:
construct TargetUris for resolved addresses

Commit:
DialogueDnsResolutionWorker::update blocks calling thread on first update

carterkozak commented on dialogue-clients/src/main/java/com/palantir/dialogue/clients/DialogueDnsResolutionWorker.java:
```diff
@@ -0,0 +1,104 @@
+                // check for updates to scb state first
+                ServicesConfigBlock updatedInputState = updatesQueue.poll(5, TimeUnit.SECONDS);
+                doUpdate(updatedInputState);
+            } catch (InterruptedException e) {
```
When an InterruptedException is thrown, the thread interrupted state is reset back to false, so the check on line 79 will not end up setting `shutdownRequested` to `true`.

bjlaub replied:
doh, tyvm. fixed

Commit:
dont check Thread#isInterrupted when exiting loop

Commit:
big todo

bjlaub commented on dialogue-clients/src/main/java/com/palantir/dialogue/clients/DialogueClients.java:
```diff
@@ -47,6 +47,19 @@
 public final class DialogueClients {
 
     public static ReloadingFactory create(Refreshable<ServicesConfigBlock> scb) {
+        // TODO(blaub): move the logic from RelaodingClientFactory ctor that launches a DialogueDnsResolutionWorker
```
@carterkozak i suspect we'll want to do something like:
- change `ReloadingParams` to hold a `Refreshable<ServicesConfigBlockWIthResolvedHosts>` instead
- create the dns worker thread here, and have it `subscribe` to the `scb` parameter
- make the thread output updates to the `Refreshable<ServicesConfigBlockWIthResolvedHosts>` on the `ReloadingParams` instance we create

open question is what to do about the `reloading()` method on `ReloadingFactory`, which comes from `ConjureClients.ToReloadingFactory` interface and creates a new factory with params that set a different ServicesConfigBlock refreshable. Perhaps call `dispose()` on the original subscription and re-wire everything to the new one without spawning a new thread?

Commit:
Ckozak/dns resolution worker (#2146)

* ReloadingFactory service extraction is robust to malformed inputs

A single bad input shouldn't prevent us from parsing other services.
Worth adding some tests as well.

* DialogueDnsResolutionWorker robustness and blocking updates

DialogueDnsResolutionWorker updates as much as possible even
when some URIs cannot be parsed.

* ReloadingClientFactory uses the correct DNS resolver
